### PR TITLE
feat(stream): support arrangement backfill frontend

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Refactor `StreamTableScan` to support arrangement backfill.
- We now pass `TableCatalog` instead of `TableDesc` around in frontend, since `TableCatalog` is a superset of `TableDesc`, and is required for arrangement backfill.
- Good enough to pass the existing test suite.
- Fix some state persistence logic of arrangement backfill.
- Refactor state table's catalog to support `output_indices`.
- Introduce a new interface for state table, so it can just read `output_indices`, similar to `StorageTable`.

Before merging:
- [x] Disable arrangement backfill by default

### Further work

- Replication is not supported yet. It requires changes to the dispatcher logic (currently all backfill is hardcoded to be no-shuffle.). It also requires further changes to the state table interface. To avoid complexity in review we will add it separately.
- Optimize state logic. I think it can be further improved.
- Test scale-in / scale-out once replication supported.
- Support row_count persistence.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/catalog/table_catalog.rs [hunk 1]
```diff
@@ -487,6 +490,29 @@ impl TableCatalog {
...
+        //     if self.value_indices.contains(output_idx) {
+        //         output_value_indices.push(*output_idx);
+        //     }
+        // }
+        Self {
+            // value_indices: output_value_indices,
+            output_indices,
+            ..self
+        }
+    }
```
Comment (kwannoel) [inline src/frontend/src/catalog/table_catalog.rs:515]: We won't do replication in this PR, since that still requires several changes. See PR description for more details.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs [hunk 2]
```diff
@@ -247,7 +262,7 @@ impl Scan {
         Self::new_inner(
             table_name,
             output_col_idx,
-            table_desc,
```
Comment (kwannoel) [inline src/frontend/src/optimizer/plan_node/generic/scan.rs:250]: We can have everything reference `table_catalog`, but that requires a fair amount of refactoring. So do that in separate PR.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs [hunk 3]
```diff
@@ -276,18 +291,21 @@ impl Scan {
...
         let mut required_col_idx = output_col_idx.clone();
-        let predicate_col_idx = predicate.collect_input_refs(table_desc.columns.len());
+        let predicate_col_idx = predicate.collect_input_refs(table_catalog.columns().len());
         predicate_col_idx.ones().for_each(|idx| {
             if !required_col_idx.contains(&idx) {
                 required_col_idx.push(idx);
             }
         });
 
+        let table_desc = Rc::new(table_catalog.table_desc());
```
Comment (kwannoel) [inline src/frontend/src/optimizer/plan_node/generic/scan.rs:301]: We can have everything reference `table_catalog`, but that requires a fair amount of refactoring. So do that in separate PR.

File: src/meta/src/stream/stream_graph/actor.rs [hunk 4]
```diff
@@ -173,6 +173,7 @@ impl ActorBuilder {
                     downstream_fragment_id: self.fragment_id,
                 }];
 
+                // FIXME: This may not hold for Arrangement Backfill.
```
Comment (kwannoel) [inline src/meta/src/stream/stream_graph/actor.rs:176]:
```suggestion
                // FIXME(kwannoel): This may not hold for Arrangement Backfill.
```

File: e2e_test/streaming/watermark.slt [hunk 5]
```diff
@@ -20,6 +20,7 @@ insert into t values ('2023-05-06 16:51:00', 1), ('2023-05-06 16:51:00', 2), ('2
 statement ok
 insert into t values ('2023-05-06 16:56:01', 1);
 
+# FIXME(kwannoel): Seems like this is flaky
```
Comment (kwannoel) [inline e2e_test/streaming/watermark.slt:23]:
```suggestion
```

File: src/stream/src/executor/backfill/arrangement_backfill.rs [hunk 6]
```diff
@@ -49,15 +50,15 @@ use crate::task::{ActorId, CreateMviewProgress};
...
     /// Upstream table
-    upstream_table: ReplicatedStateTable<S>,
+    upstream_table: ReplicatedStateTable<S, SD>,
 
     /// Upstream with the same schema with the upstream table.
     upstream: BoxedExecutor,
 
     /// Internal state table for persisting state of backfill state.
-    state_table: ReplicatedStateTable<S>,
+    state_table: StateTable<S>,
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply src/stream/src/executor/backfill/arrangement_backfill.rs:61]: > However, if there is no replication, how does it read the non-ckpt data from this state table in arrangement backfill?

This is intentional. Arrangement backfill in this PR is basically `no_shuffle_backfill` but with vnode level snapshot read.

In future PR I will add replication and amend the dispatcher shuffling logic for it. Not adding in this PR, because it will make it harder to review.

May take a look at `#future-work` in PR description for next steps.

**Comment by kwannoel** [comment]
Tests passed.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs [hunk 7]
```diff
@@ -37,9 +39,10 @@ pub struct Scan {
     /// Include `output_col_idx` and columns required in `predicate`
     pub required_col_idx: Vec<usize>,
     pub output_col_idx: Vec<usize>,
-    /// Descriptor of the table
-    pub table_desc: Rc<TableDesc>,
     /// Descriptors of all indexes on this table
+    /// Table Catalog of the upstream table that the descriptor is derived from.
+    pub table_catalog: Arc<TableCatalog>,
+    // Descriptors of all indexes on this table
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/generic/scan.rs:45]: Comments seem unmatched.

File: src/frontend/src/catalog/table_catalog.rs [hunk 8]
```diff
@@ -155,6 +154,9 @@ pub struct TableCatalog {
 
     /// description of table, set by `comment on`.
     pub description: Option<String>,
+
+    /// Output indices, only used by replicated state table.
+    pub output_indices: Vec<usize>,
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply src/frontend/src/catalog/table_catalog.rs:159]: > If the output_indices is only used by the replicated state table, what about adding them later?
We still need it here. Backfill reads from storage table with output indices. Similarly here, we also need to read with output indices. We can either include that field in backfill or state table. But since replicated state table will use it, I think it's better to add it in state table and avoid a refactor later.

> BTW, table catalog with an ouput_indices seems a little weird. Can we maintain this field in the stream scan proto?
I prefer not to. Main reason is that replicated state table will be constructed from table catalog.
We could pass it in from stream scan, but IMO state table's meta data (e.g. output_indices) should be entirely contained in table catalog.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs [hunk 9]
```diff
@@ -37,9 +39,10 @@ pub struct Scan {
     /// Include `output_col_idx` and columns required in `predicate`
     pub required_col_idx: Vec<usize>,
     pub output_col_idx: Vec<usize>,
-    /// Descriptor of the table
-    pub table_desc: Rc<TableDesc>,
     /// Descriptors of all indexes on this table
```
Comment (kwannoel) [inline src/frontend/src/optimizer/plan_node/generic/scan.rs:42]:
```suggestion
```

File: src/frontend/src/optimizer/plan_node/stream_table_scan.rs [hunk 10]
```diff
@@ -282,16 +302,19 @@ impl StreamTableScan {
...
+        let upstream_table_catalog = self
+            .get_upstream_state_table()
+            .clone()
+            .into_replicated(output_indices_origin);
+
+        let arrangement_table = if self.stream_scan_type == StreamScanType::ArrangementBackfill {
+            Some(upstream_table_catalog.to_internal_table_prost())
+        } else {
+            None
         };
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/stream_table_scan.rs:316]: According to these lines, `output_indices_origin` is used by `ArrangementBackfill` only. What do you mean it is also used by the no shuffle backfill?

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_table_scan.rs:316]: I mean storage table has a similar concept as well:
```
    /// Create a  [`StorageTableInner`] given a complete set of `columns` and a partial
    /// set of `column_ids`. The output will only contains columns with the given ids in the same
    /// order.
    #[allow(clippy::too_many_arguments)]
    pub fn new_partial(
        store: S,
        table_id: TableId,
        table_columns: Vec<ColumnDesc>,
        column_ids: Vec<ColumnId>,
        order_types: Vec<OrderType>,
        pk_indices: Vec<usize>,
        distribution: Distribution,
        table_options: TableOption,
        value_indices: Vec<usize>,
        read_prefix_len_hint: usize,
        versioned: bool,
    ) -> Self {
```

Where it uses `column_ids` to only read a specific set of columns.

The usage is similar to output_indices.

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_table_scan.rs:316]: Which is what `no_shuffle_backfill` uses

Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/stream_table_scan.rs:316]: My point is that `TableCatalog` is a struct used more than just for arrangement backfill. Adding an additional field just for arrangement backfill seems overkill. As you mentioned, for non-shuffle backfill `StreamScanNode` has a separate field `upstream_column_ids` to represent the ouput_indices, can we do that for the arrangement backfill in the same way? That would make `StreamScanNode` meta data more consistent between arrangement and no shuffle backfill.

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_table_scan.rs:316]: > My point is that `TableCatalog` is a struct used more than just for arrangement backfill. Adding an additional field just for arrangement backfill seems overkill. As you mentioned, for non-shuffle backfill `StreamScanNode` has a separate field `upstream_column_ids` to represent the ouput_indices, can we do that for the arrangement backfill in the same way? That would make `StreamScanNode` meta data more consistent between arrangement and no shuffle backfill. 

Alright, let me try to unify them.

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_table_scan.rs]: Including output_indices in the table catalog made it a lot more complicated than it had to be. 

I have now fixed it. Thanks for the great suggestion.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs [hunk 11]
```diff
@@ -169,12 +172,23 @@ impl Scan {
         ids
     }
 
+    pub fn output_and_pk_column_indices(&self) -> Vec<usize> {
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/generic/scan.rs:175]: unused function.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs [hunk 12]
```diff
@@ -37,8 +39,8 @@ pub struct Scan {
     /// Include `output_col_idx` and columns required in `predicate`
     pub required_col_idx: Vec<usize>,
     pub output_col_idx: Vec<usize>,
-    /// Descriptor of the table
-    pub table_desc: Rc<TableDesc>,
+    /// Table Catalog of the upstream table that the descriptor is derived from.
+    pub table_catalog: Arc<TableCatalog>,
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: Which fields used in the arrangement backfill upstream table exist only in `TableCatalog ` but not in `TableDesc`?

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: We use the function `from_table_catalog` to construct state table. Let me see if it's possible to derive it from `TableDesc`.

File: proto/stream_plan.proto [hunk 13]
```diff
@@ -482,6 +482,9 @@ enum StreamScanType {
 
   // ChainExecutor with upstream_only = true
   STREAM_SCAN_TYPE_UPSTREAM_ONLY = 4;
+
+  // ARRANGEMENT_BACKFILL corresponds to the `ArrangementBackfillExecutor`.
+  ARRANGEMENT_BACKFILL = 5;
```
Comment (fuyufjh) [inline proto/stream_plan.proto:487]: The prefix is only used as a namespace, and will be removed by the rust library (prost).

```suggestion
  STREAM_SCAN_TYPE_ARRANGEMENT_BACKFILL = 5;
```

File: src/frontend/src/optimizer/plan_node/generic/scan.rs (hunk 12 above)
Comment (fuyufjh) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: related https://github.com/risingwavelabs/risingwave/issues/13146#issuecomment-1797674824

File: src/stream/src/executor/backfill/utils.rs [hunk 14]
```diff
@@ -79,32 +89,147 @@ impl BackfillState {
...
+        }
+        let new_state = self.inner.get(vnode).unwrap().current_state().clone();
+        let new_encoded_state = match new_state {
+            BackfillProgressPerVnode::NotStarted => unreachable!(),
+            BackfillProgressPerVnode::InProgress(current_pos) => {
+                let mut encoded_state = vec![None; current_pos.len() + METADATA_STATE_LEN];
+                encoded_state[0] = Some(vnode.to_scalar().into());
+                encoded_state[1..current_pos.len() + 1].clone_from_slice(current_pos.as_inner());
+                encoded_state[current_pos.len() + 1] = Some(false.into());
+                encoded_state[current_pos.len() + 2] = Some(0i64.into());
```
Comment (chenzl25) [inline src/stream/src/executor/backfill/utils.rs:130]: The `row_count` field is always zero? It seems we should pass `total_snapshot_processed_rows` here.

File: src/stream/src/executor/backfill/utils.rs [hunk 15]
```diff
@@ -79,32 +89,147 @@ impl BackfillState {
...
+            }
+        };
+        *current_state = BackfillProgressPerVnode::Completed(new_pos);
+    }
+
+    /// If can commit, return state to be committed.
+    fn get_commit_state(&self, vnode: &VirtualNode) -> Option<(Option<Vec<Datum>>, Vec<Datum>)> {
+        if !self.can_commit(vnode) {
+            return None;
+        }
```
Comment (chenzl25) [inline src/stream/src/executor/backfill/utils.rs:121]: Moving this code outside the `get_commit_state` would make it more straightforward.

File: src/stream/src/executor/backfill/utils.rs [hunk 16]
```diff
@@ -79,32 +89,147 @@ impl BackfillState {
...
+                encoded_state[committed_pos.len() + 1] = Some(true.into());
+                encoded_state[committed_pos.len() + 2] = Some(0i64.into());
+                Some(encoded_state)
+            }
+        };
+        Some((old_encoded_state, new_encoded_state))
+    }
+
+    // TODO: We can add a committed flag to speed up this check.
+    fn can_commit(&self, vnode: &VirtualNode) -> bool {
```
Comment (chenzl25) [inline src/stream/src/executor/backfill/utils.rs:168]: `need_commit` seems a more suitable function name.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs (hunk 12 above)
Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: > Which fields used in the arrangement backfill upstream table exist only in TableCatalog but not in TableDesc?

- table option (can technically be passed in to table desc, since now table desc contains the sole field (retention seconds))
- vnode_col_index, just a single field, I suppose we can also include it in table desc as well.

So I think we could use `TableDesc` too.

The rest of the fields can be derived in some way, although they may not be present in `TableDesc`.

Is there some reason to prefer `TableDesc` vs `TableCatalog`? 🤔 My thought process is that it's the least error prone to just clone the upstream catalog.

Is it just to reduce the refactoring changes to frontend?

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: Hmm that will require using and modifying the proto definition of `StorageTableDesc`, which does not seem preferred... That should be reserved for `StorageTable`.

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: I will open a separate PR to do some refactoring. Seems like `TableCatalog` can be used everywhere in frontend instead, so we shouldn't need to maintain a separate interface, just for `TableDesc`.

File: src/stream/src/executor/backfill/utils.rs (hunk 14 above)
Comment (kwannoel) [reply src/stream/src/executor/backfill/utils.rs:130]: I will refactor it separately. https://github.com/risingwavelabs/risingwave/issues/10225.

Requires changes to the vnode level state.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs (hunk 12 above)
Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: Separated `arrangement backfill`'s proto definition from the stream proto definition.

We can freely change it ourselves, it won't be exposed to users.

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: Still use `TableCatalog` (via `catalog.Table`) to construct replicated state table.

In the future we can do further refactoring: https://github.com/risingwavelabs/risingwave/issues/13868

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs:43]: Seems like it will cause a lot of breakage, since several areas of code expect `StreamScan`. I will parameterize the upstream table instead using `oneof`.

File: src/frontend/src/optimizer/plan_node/generic/scan.rs [hunk 17]
```diff
@@ -52,6 +54,7 @@ pub struct Scan {
     #[educe(PartialEq(ignore))]
     #[educe(Hash(ignore))]
     pub ctx: OptimizerContextRef,
+    pub table_desc: Rc<TableDesc>,
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/generic/scan.rs]: See comment: https://github.com/risingwavelabs/risingwave/pull/13622/files#r1405675502

I have created an issue for it, https://github.com/risingwavelabs/risingwave/issues/13874.
This field will be removed later.

**Comment by kwannoel** [comment]
Tests passed, including latest changes for rate limit.

File: src/stream/src/executor/backfill/arrangement_backfill.rs [hunk 18]
```diff
@@ -159,13 +175,19 @@ where
...
+                        backfill_state.clone(), // FIXME(kwannoel): use mutable reference instead.
+                        &mut builders,
+                    );
+                    pin_mut!(snapshot);
+                    let next = snapshot.try_next().await?;
+                    next.is_none()
+                };
+                let builder_is_empty = builders.iter().all(|b| b.is_empty());
+                builders.iter_mut().for_each(|b| b.clear());
+                snapshot_is_empty && builder_is_empty
```
Comment (chenzl25) [inline src/stream/src/executor/backfill/arrangement_backfill.rs:190]: Maybe we can get rid of this empty snapshot check? I remember we used to remove it from the no shuffle backfill.

**Review by chenzl25 — APPROVED** [review]
Rest LGTM. Thank you for your effort!

File: src/stream/src/executor/backfill/arrangement_backfill.rs (hunk 18 above)
Comment (kwannoel) [reply src/stream/src/executor/backfill/arrangement_backfill.rs]: Fixed

**Comment by kwannoel** [comment]
`expect-test` used in `src/storage/src/row_serde/mod.rs` after this PR to test some utility function.

**Comment by kwannoel** [comment]
Session config has been reverted.

**Comment by kwannoel** [comment]
Can someone help to approve the code-owners check? For the cargo lock file.

**Review by fuyufjh — APPROVED** [review]

File: src/stream/src/common/table/state_table.rs [hunk 19]
```diff
@@ -1131,6 +1215,25 @@ where
         ))
     }
 
+    pub async fn iter_with_vnode_and_output_indices(
```
Comment (BugenZhao) [inline src/stream/src/common/table/state_table.rs:1218]: I'm worrying if this can be error-prone: one creates a state table with output indices but forget to iter with this method, or vice versa.

Is it okay to make `output_indices` an `Option`? Then we can always project with it if present, or do nothing if absent through the same entrance.

File: src/stream/src/common/table/state_table.rs [hunk 20]
```diff
@@ -138,13 +142,23 @@ pub struct StateTableInner<
...
+
+    /// Data Types
+    /// We will need to use to build data chunks from state table rows.
+    data_types: Vec<DataType>,
+
+    /// Output indices
+    /// Used for:
+    /// 1. Computing output_value_indices to ser/de replicated rows.
+    /// 2. Computing output pk indices to used them for backfill state.
+    output_indices: Vec<usize>,
```
Comment (BugenZhao) [inline src/stream/src/common/table/state_table.rs:154]: How will we handle `output_indices` on the write path?

File: src/stream/src/common/table/state_table.rs [hunk 21]
```diff
@@ -138,13 +142,23 @@ pub struct StateTableInner<
 
     /// Watermark cache
     watermark_cache: StateTableWatermarkCache,
+
+    /// Data Types
+    /// We will need to use to build data chunks from state table rows.
+    data_types: Vec<DataType>,
```
Comment (BugenZhao) [inline src/stream/src/common/table/state_table.rs:148]: I guess it should be accessible within the deserializer from `SD`?

File: src/stream/src/common/table/state_table.rs (hunk 20 above)
Comment (kwannoel) [reply src/stream/src/common/table/state_table.rs:154]: For now no replication, subsequent PR will handle that.

File: src/stream/src/common/table/state_table.rs (hunk 21 above)
Comment (kwannoel) [reply src/stream/src/common/table/state_table.rs:148]: Yes but seems a little strange to expose it via the deserializer.
